### PR TITLE
2.3 Assignments Matt Mastin

### PR DESCRIPTION
**Pull request by mmastin**

### CODE REVIEW

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Matthew Mastin
## Assignment: log_linear_regression


# Part 1. Importation

***
**What went well:**

-Good use of glob to find all the pathnames matching a specified patter:

```python
from glob import glob
import pandas as pd

for path in glob('competition_data/*.csv'):
  df = pd.read_csv(path)
  print(path, df.shape)

```



# Splitting Data

## Imports and MSE
-Great job with implementing train_test_split and nice random state :) : 

```python
# splitting data into train and validation sets

from sklearn.model_selection import train_test_split
trainval_tube_assemblies, val_tube_assemblies = train_test_split(trainval_tube_assemblies,
                                                                random_state=42)
```
**What could be added**

- baseline over 90 and with clean code and comments :

  
```python

target = 'cost'
y_train = train[target]
y_val = val[target]
y_pred = np.full_like(y_val, fill_value=y_train.mean()) #this numpy method creates a dataframe with the shape of a given dataframe, here we are filling it with mean

print('Validation RMSLE, Mean Baseline:', rmsle(y_val, y_pred))

#Mean Baseline: 0.94


```

# Forest Regression

-  Random Forest

```python
import matplotlib.pyplot as plt

rf = pipeline.named_steps['randomforestregressor']
importances = pd.Series(rf.feature_importances_, df_train.columns)
importances.sort_values().plot.barh();

```


**What could be added**
# Making a Pipeline 

- Importation of RFR :

```python


from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
model.fit(X_train, y_train)
r2 = model.score(X_val, y_val)
print('Validation R^2:', r2)

#Validation R^2: 0.08
```

- Importation of sklearn make_pipeline : 

```python

import category_encoders as ce
from sklearn.pipeline import make_pipeline

pipeline = make_pipeline(
  ce.OrdinalEncoder(),
    RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
)

pipeline.fit(X_train, y_train_log)
y_pred_log = pipeline.predict(X_val)
rmse(y_val_log, y_pred_log)


```

# What to work on
- [x] Finish your assignment
- [x] Show up to stand up!
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Matthew Mastin
## Assignment: Gradient Boosting 


# Part 1. Importation/Unzipping

***
**What went well:**

-Good use of glob to find all the pathnames matching a specified patter:

```python
with ZipFile(file_name, 'r') as zip:
    zip.printdir()
    print('extracting all files now...')
    zip.extractall()
    print('Done')


!unzip data.zip

```



# Splitting Data

- Good split here and valid syntax:

```python
trainval_tube_assemblies = trainval['tube_assembly_id'].unique()
train_tube_assemblies, val_tube_assemblies = train_test_split(trainval_tube_assemblies, random_state=42)
train = trainval[trainval['tube_assembly_id'].isin(train_tube_assemblies)]
val = trainval[trainval['tube_assembly_id'].isin(val_tube_assemblies)]
```

## Imports and MSE
-Great job finding RMSE with XGBRegressor : 

```python
# splitting data into train and validation sets



# %%time
from xgboost import XGBRegressor
pipeline = make_pipeline(ce.OrdinalEncoder(),
                        XGBRegressor(n_estimators=1000, n_jobs=-1))
pipeline.fit(X_train, y_train_log)
y_pred_log = pipeline.predict(X_val)
print('RMSE error for Xgbr', rmse(y_val_log, y_pred_log))




```
**What could be added**

- Done with sklearn:

  
```python

from sklearn.model_selection import train_test_split
train_tube_assemblies, val_tube_assemblies = train_test_split(
  trainval_tube_assemblies, random_state=42)


```

# Forest Regression

-  Random Forest

```python
import matplotlib.pyplot as plt

rf = pipeline.named_steps['randomforestregressor']
importances = pd.Series(rf.feature_importances_, df_train.columns)
importances.sort_values().plot.barh();

```


**What could be added**
# Making a Pipeline 

```python
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
model.fit(X_train, y_train)
r2 = model.score(X_val, y_val)
print('Validation R^2:', r2)

#Validation R^2: 0.08
```

- Importation of sklearn make_pipeline : 

```python

import category_encoders as ce
from sklearn.pipeline import make_pipeline

pipeline = make_pipeline(
  ce.OrdinalEncoder(),
    RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
)

pipeline.fit(X_train, y_train_log)
y_pred_log = pipeline.predict(X_val)
rmse(y_val_log, y_pred_log)


```

# What to work on
- [x] Complete Random Forest section
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Matthew Mastin
## Assignment: Partial Dependence  


# Part 1. Importation/Unzipping

***
**What went well:**

-Great job in the stages of preprocessing:

```python
assembly_components = materials.melt(id_vars='tube_assembly_id',
                                    value_vars=[f'component_id_{n}' for n in range(1,9)])
assembly_components = (assembly_components.sort_values(by='tube_assembly_id')
                      .dropna()
                      .rename(columns={'value': 'component_id'}))
assembly_components.head()

```



# RMSE

- Great job using xgboost to encode your X and Y train data:

```python
import category_encoders as ce
from xgboost import XGBRegressor
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

encoder = ce.OrdinalEncoder()
X_train_encoded = encoder.fit_transform(X_train)
X_val_encoded = encoder.transform(X_val)
eval_set = [(X_train_encoded, y_train_log),
           (X_val_encoded, y_val_log)]
model = XGBRegressor(n_estimators=2000, n_jobs=-1)
model.fit(X_train_encoded, y_train_log,
         eval_set=eval_set, eval_metric='rmse', early_stopping_rounds=50)
```

## Graphing
- Nice pretty plot with pdpbox : 

```python
from pdpbox.pdp import pdp_interact, pdp_interact_plot

features = ['month', 'year']

interaction = pdp_interact(model=model,
                          dataset=X_val_encoded,
                          model_features=X_val_encoded.columns,
                          features=features)

pdp_interact_plot(interaction, plot_type='grid', feature_names=features);


```

# What to work on
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Matthew Mastin
## Assignment: Hyperparameter Optimization 


# Part 1. Importation/Unzipping

***
**What went well:**

-Nice job, you can also use glob here :) :

```python
train = pd.read_csv('competition_data/train_set.csv')
test = pd.read_csv('competition_data/test_set.csv')
tube = pd.read_csv('competition_data/tube.csv')
materials = pd.read_csv('competition_data/bill_of_materials.csv')
components = pd.read_csv('competition_data/components.csv')

comp = pd.concat((pd.read_csv(path) for path in glob('competition_data/comp_*.csv')), sort=False)

```
-Great data cleaning:

```python
assembly_components = materials.melt(id_vars='tube_assembly_id',
                                    value_vars=[f'component_id_{n}' for n in range(1,9)])

assembly_components = (assembly_components
                      .sort_values(by='tube_assembly_id')
                      .dropna()
                      .rename(columns={'value': 'component_id'}))

assembly_component_types = assembly_components.merge(components, how='left')

```


# RMSE

- Great job using xgboost to encode your X and Y train data:

```python
import category_encoders as ce
from xgboost import XGBRegressor
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

encoder = ce.OrdinalEncoder()
X_train_encoded = encoder.fit_transform(X_train)
X_val_encoded = encoder.transform(X_val)
eval_set = [(X_train_encoded, y_train_log),
           (X_val_encoded, y_val_log)]
model = XGBRegressor(n_estimators=2000, n_jobs=-1)
model.fit(X_train_encoded, y_train_log,
         eval_set=eval_set, eval_metric='rmse', early_stopping_rounds=50)
```

## Grid Search/Result
- Good use of GridSearchCV and creating an output with the n_estimators given : 

```python
%%time
# Trying GridSearch as stretch goal

from sklearn.model_selection import GridSearchCV

param_grid = {'xgbregressor__n_estimators': (540,570),
            'xgbregressor__max_depth': (5,6)}

gs = GridSearchCV(estimator=pipeline,
                     param_grid=param_grid,
                     scoring='neg_mean_squared_error',
                     n_jobs=-1,
                     cv=3,
                     verbose=3,
                      refit=True,
)

gs.fit(X_train, y_train_log);

print('best hyperparamters', search.best_params_)
print('cross val rmsle', np.sqrt(-search.best_score_))

#best hyperparamters {'xgbregressor__max_depth': 5, 'xgbregressor__n_estimators': 545}
#cross val rmsle 0.31088637924275075


```

# What to work on
- [x] Try to read code backwards (makes for less complicated coding)
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Matthew Mastin
## Sprint Challenge: Regression 2 


# Part 1. Wrangle Data

***
**What went well:**

-Nice merge and use of pandas to merge the two datasets:

```python
# Merging train and test with store
trainstore = train.merge(store, how='left')
print(trainstore.shape)
trainstore.head()


```

- Great job using .describe and making sure to exclude the number in trainstore: 

```python 
trainstore.describe()
trainstore.describe(exclude='number')


```

**What could be better**


- Plot could be easily applied using seaborn:

```python 
%matplotlib inline
import seaborn as sns
sns.distplot(y_train);


sns.distplot(y_train_log);

```



# Fit and validate your model

- Good cross val score using xgboost and category encoding: 

```python
import category_encoders as ce
from xgboost import XGBRegressor
from sklearn.pipeline import make_pipeline
from sklearn.metrics import mean_squared_error, mean_squared_log_error
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

def rmse(y_true, y_pred):
    return np.sqrt(mean_squared_error(y_true, y_pred))

# Error on the model.fit line with this
# encoder = ce.OrdinalEncoder()
# X_train_encoded = encoder.fit_transform(X_train)
# X_val_encoded = encoder.transform(X_val)

# model = XGBRegressor(n_estimators=700, n_jobs=-1, random_state=24)
# model.fit(X_train_encoded, y_train_log)
# y_pred_log = model.predict(X_val_encoded)
# print('RMSE Score', rmse(y_val_log, y_pred_log))

pipeline = make_pipeline(ce.OrdinalEncoder(),
                        XGBRegressor(n_estimators=1000, n_jobs=-1, random_state=24))
pipeline.fit(X_train, y_train_log)
y_pred_log = pipeline.predict(X_val)
print('RMSE for XGBR', rmse(y_val_log, y_pred_log))
#RMSE for XGBR 0.05751556924720625

```

**What could be better**

- Making a pipeline with k_means and XGBRegressor is an ideal solution as well (slightly better score):

```python
%%time
import category_encoders as ce
from sklearn.model_selection import cross_val_score
from sklearn.pipeline import make_pipeline
#from conda install scikit-learnxgboost import XGBRegressor

pipeline = make_pipeline(ce.OrdinalEncoder(), XGBRegressor(random_state=42, n_jobs=-1))

k = 3
groups = train['Store']
scores = cross_val_score(pipeline, X_train, y_train_log, cv=k, n_jobs=-1, 
                         scoring='neg_mean_squared_error', groups=groups)
print(f'RMSLE for {k} folds:', np.sqrt(-scores))

```


# Plot model interpretation visualizations

- Great job taking in pdpbox and getting a feel for the features within your data:

```python
import eli5
from eli5.sklearn import PermutationImportance
from pdpbox.pdp import pdp_isolate, pdp_plot

# PDP Plot, 1 Feature

feature = 'DayOfWeek'

isolated = pdp_isolate(model=model,
                      dataset=X_val_encoded,
                      model_features=X_val_encoded.columns,
                      feature=feature)
pdp_plot(isolated, feature_name=feature)




```
**What could be better**

- pdp is a great way to see optimal category:

```python
from pdpbox.pdp import pdp_isolate, pdp_plot

feature = 'Customers'

isolated = pdp_isolate(
    model=model, 
    dataset=X_val_encoded, 
    model_features=X_val_encoded.columns, 
    feature=feature, 
    num_grid_points=20 
)

pdp_plot(isolated, feature_name=feature);
```


- Another example:

```python
from pdpbox.pdp import pdp_interact, pdp_interact_plot

features = ['Customers', 'CompetitionDistance']

interaction = pdp_interact(
    model=model, 
    dataset=X_val_encoded, 
    model_features=X_val_encoded.columns, 
    features=features
)

pdp_interact_plot(interaction, plot_type='grid', feature_names=features);
```


# What to work on
- [x] Try to read code backwards (makes for less complicated coding)
- [x] Come to PM for questions or help
